plugins/arm-lifter/armng_env.ml

```diff
@@ -0,0 +1,101 @@
+open Bap_core_theory
+open Base
+open KB.Syntax
+
+module Env = struct
+  type value
+  type byte
+  type reg_single
+  type half_byte
+  type half_word
+  type bit_val
+  type double_word
+
+  let bit : Theory.Bool.t Theory.Value.sort =
+    Theory.Bool.t
+
+  let bit_val : bit_val Theory.Bitv.t Theory.Value.sort =
+    Theory.Bitv.define 1
+
+  let double_word : double_word Theory.Bitv.t Theory.Value.sort =
+    Theory.Bitv.define 64
+
+  (** grps are defined by 3-bit indices *)
+  let reg_single : reg_single Theory.Bitv.t Theory.Value.sort =
+    Theory.Bitv.define 3
+
+  let half_word : half_word Theory.Bitv.t Theory.Value.sort =
+    Theory.Bitv.define 16
+
+  let value : value Theory.Bitv.t Theory.Value.sort =
+    Theory.Bitv.define 32
+
+  let half_byte : half_byte Theory.Bitv.t Theory.Value.sort =
+    Theory.Bitv.define 4
+
+  let byte : byte Theory.Bitv.t Theory.Value.sort =
+    Theory.Bitv.define 8
+
+  let heap_type = Theory.Mem.define value byte
+
+  let memory = Theory.Var.define heap_type "mem"
+
+  (** define grps *)
+  let r0 = Theory.Var.define value "r0"
+  let r1 = Theory.Var.define value "r1"
+  let r2 = Theory.Var.define value "r2"
+  let r3 = Theory.Var.define value "r3"
+  let r4 = Theory.Var.define value "r4"
+  let r5 = Theory.Var.define value "r5"
+  let r6 = Theory.Var.define value "r6"
+  let r7 = Theory.Var.define value "r7"
+  let r8 = Theory.Var.define value "r8"
+  let r9 = Theory.Var.define value "r9"
+  let r10 = Theory.Var.define value "r10"
+  let r11 = Theory.Var.define value "r11"
+  let r12 = Theory.Var.define value "r12"
+
+  let lr = Theory.Var.define value "lr"
+  let pc = Theory.Var.define value "pc"
```

Comment:
The PC register is not really a register so it should be removed from here. All accesses to the PC register should be resolved to static constants that are equal to the address of the current instruction + some ISA specific offset. See how it was done in all other lifters.

Reply:
For this , one of the problem I'm encountering now is that: different from BIL, Core Theory represents data effect and control effect with different polymorphic variable (`Theory.data` and `Theory.ctrl`), that could only be unified with `blk`, which seems to me to be representing a program `block` (like that of LLVM IR), leads me to the conclusion that this can only be produced once each instruction. (tell me if it's not right)

So, currently, to make the lifter sub-routines(for each instruction) smooth, I made those functions with both effects to return a `(data effect, ctrl effect)` tuple and finally resolve them with `blk`, I'm feeling comfortable with this because only a small set of instructions needs this trick and they do not much affect the over-all consistency.

However, considering that we are now abolishing the `PC` register var, and considering the fact that any instruction with full GRPs access can access `PC` before ARMv8, the **small set** previously mentioned almost extends to every single ARM instruction (a catastrophe for current implemetation).

I do think the previous approach should be abolished in this case, but I'm hesitate over how to properly resolve this, one of the thought is to define the DSL over a BiMonad to carry both effects, but I'm afraid this could complicate things.

Another minor problem is that how can we make this substitution possible in Core Theory/KB system:
```ocaml
(** Substitute PC with its value  *)
let resolve_pc mem = Stmt.map (object(self)
    inherit Stmt.mapper as super
    method! map_var var =
      if Var.(equal var CPU.pc) then
        Bil.int (CPU.addr_of_pc mem)
      else super#map_var var
  end)
```